sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -363,48 +365,125 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
+                s"The column number of the existing table $tableName" +
+                  s"(${existingSchema.catalogString}) doesn't match the data schema" +
+                  s"(${df.logicalPlan.schema.catalogString})")
+            }
+
+            if (partitioningColumns.isDefined) {
+              logWarning("append to an existing table, the specified partition columns " +
+                s"[${partitioningColumns.get.mkString(", ")}] will be ignored.")
+            }
+
+            val specifiedBucketSpec = getBucketSpec
+            if (specifiedBucketSpec.isDefined) {
+              logWarning("append to an existing table, the specified bucketing " +
+                s"(${specifiedBucketSpec.get}) will be ignored.")
+            }
+
+            // Because we are inserting into an existing table, we should respect the existing
+            // schema and adjust columns order of the given dataframe according to it.
+            val adjustedColumns = existingSchema.map(f => Column(df.resolve(f.name)))
+            df.select(adjustedColumns: _*).write.insertInto(tableIdentWithDB)
+
+          case SaveMode.Overwrite =>
+            EliminateSubqueryAliases(catalog.lookupRelation(tableIdentWithDB)) match {
+              // Only do the check if the table is a data source table
+              // (the relation is a BaseRelation).
+              case LogicalRelation(dest: BaseRelation, _, _) =>
+                // Get all input data source relations of the query.
+                val srcRelations = df.logicalPlan.collect {
+                  case LogicalRelation(src: BaseRelation, _, _) => src
+                }
+                if (srcRelations.contains(dest)) {
+                  throw new AnalysisException(
+                    s"Cannot overwrite table $tableName that is also being read from")
+                }
+              case _ => // OK
+            }
+
+            // Drop the existing table and create it again.
+            catalog.dropTable(tableIdentWithDB, ignoreIfNotExists = true, purge = false)
+            createTable(tableIdentWithDB)
+
+            // Refresh the cache of the table in the catalog.
+            catalog.refreshTable(tableIdentWithDB)
         }
 
-        val tableDesc = CatalogTable(
-          identifier = tableIdent,
-          tableType = tableType,
-          storage = storage,
-          schema = new StructType,
-          provider = Some(source),
-          partitionColumnNames = partitioningColumns.getOrElse(Nil),
-          bucketSpec = getBucketSpec
-        )
-        df.sparkSession.sessionState.executePlan(
-          CreateTable(tableDesc, mode, Some(df.logicalPlan))).toRdd
+      // If the table doesn't exist...
+      case None => createTable(tableIdentWithDB)
     }
   }
 
+  private def createTable(tableIdent: TableIdentifier): Unit = {
+    val storage = DataSource.buildStorageFormatFromOptions(extraOptions.toMap)
+    val tableType = if (storage.locationUri.isDefined) {
```

Comment:
I reverted https://github.com/apache/spark/pull/15983 here because it's not needed anymore after this refactor.